torch/csrc/jit/passes/onnx/eval_peephole.cpp

```diff
@@ -142,9 +142,63 @@ static void fuseConvBatchNorm(Block* b, ValueToParamPairMap& valsToParamsMap) {
   }
 }
 
+static void DeduplicateParameters(
+    Block* b,
+    ValueToParamPairMap& valsToParamsMap) {
+  // Parameters are considered as duplicated if their data_ptr, strides and
+  // sizes are equal.
+  std::vector<Value*> uniqueVals;
+  std::vector<size_t> inputsIndicesToRemove;
+  auto g = b->owningGraph();
+
+  auto is_same_tensor_as = [&valsToParamsMap](Value* v) {
+    return [&valsToParamsMap, v](Value* v2) {
+      if ((valsToParamsMap.find(v) == valsToParamsMap.end()) ||
+          (valsToParamsMap.find(v2) == valsToParamsMap.end())) {
+        return false;
+      }
+      auto iv1 = valsToParamsMap.find(v)->second.second;
+      auto iv2 = valsToParamsMap.find(v2)->second.second;
+      if (!iv1.isTensor() || !iv2.isTensor()) {
+        return false;
+      }
+      auto t1 = iv1.toTensor();
+      auto t2 = iv2.toTensor();
+      return t1.sizes().equals(t2.sizes()) &&
+          t1.strides().equals(t2.strides()) && (t1.data_ptr() == t2.data_ptr());
+    };
+  };
+
+  for (auto i : c10::irange(b->inputs().size())) {
+    auto v = g->inputs().at(i);
+    if (valsToParamsMap.find(v) == valsToParamsMap.end()) {
+      // Skip model inputs
+      continue;
+    }
+    auto it = std::find_if(
+        uniqueVals.begin(), uniqueVals.end(), is_same_tensor_as(v));
+    if (it == uniqueVals.end()) {
+      uniqueVals.emplace_back(v);
+    } else {
+      inputsIndicesToRemove.emplace_back(i);
+      v->replaceAllUsesWith(*it);
+    }
+  }
+  for (auto it = inputsIndicesToRemove.rbegin();
+       it != inputsIndicesToRemove.rend();
+       ++it) {
+    g->eraseInput(*it);
+  }
+}
+
 void EvalPeepholeONNX(Block* b, ParamMap& paramsDict) {
+  auto opset_version = GetOpsetVersion();
   auto valsToParamsMap = buildValueToParamsMap(b, paramsDict);
   fuseConvBatchNorm(b, valsToParamsMap);
+  if (opset_version >= OPSET_VERSION_15) {
+    // Apply de-duplication after opset 15 for backward compatibility.
```

Comment:
This is open to discussion. Not related to ONNX release, but we are changing default behavior of exporter regarding initializers, so I'm wondering if we should only apply it on newer opsets for BC purpose. This applies for other future passes of similar types as well. Would love to hear your thoughts on this.

Reply:
I think unless we have reason to believe this will not work on older runtimes, we should not guard the behavior.
I suggest we test this against Caffe2 using test_pytorch_onnx_caffe2.py. If we want to be extra careful you could also install an older version of ORT and run test_pytorch_onnx_onnxruntime.py.
If it passes both of those, then I think it's likely that we don't need to guard this behavior.

WDYT?

Reply:
Ok we can give a try. For general usage I'm not too worried. If there are issues they are considered bugs and we can resolve them. I'm slightly worried against user custom behaviors on top of initializer/parameters.

Reply:
The `test_operator.py` result is one example, the change in parameters may appear confusing for users. For better readability I updated the logic to insert identity node instead of directly replacing initializers.